File: code/backend/app/usuarios/router/usuario.py
from datetime import date
from fastapi import APIRouter, Depends,HTTPException,status
from sqlalchemy.orm import Session
from database import get_db
from app.usuarios.model.usuario import Usuario
from app.usuarios.schema.usuario import RegistrarUsuario,LoginUsuario,MostrarUsuario,TokenResponse
from app.core.security import hash_contrasena,verificar_contrasena
from app.core.auth import crear_token 
import re

router = APIRouter(tags=["usuarios"])

@router.post("/registrar",response_model=TokenResponse,status_code=status.HTTP_201_CREATED)
def registrar_usuario(datos:RegistrarUsuario,db:Session=Depends(get_db)):
    if datos.correo_electronico is None or datos.contrasena is None or datos.fecha_nacimiento is None or datos.profesion is None or datos.clinica is None:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST,detail="Todos los campos son obligatorios")
    if len(datos.contrasena) < 8:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST,detail="La contraseña debe tener al menos 8 caracteres")
    if not re.search(r"[A-Z]", datos.contrasena) or not re.search(r"[a-z]", datos.contrasena) or not re.search(r"\d", datos.contrasena):
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST,detail="La contraseña debe contener al menos una letra mayúscula, una letra minúscula y un número")

    if datos.fecha_nacimiento >= date.today():
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST,detail="La fecha de nacimiento debe ser anterior a la fecha actual")


    usuario_existe = db.query(Usuario).filter(Usuario.correo_electronico == datos.correo_electronico).first()
    if usuario_existe is not None:
        raise HTTPException(status_code=status.HTTP_400_BAD_REQUEST,detail="El correo electronico ya esta registrado")
    try:
        nuevo_usuario = Usuario(
        correo_electronico = datos.correo_electronico,
        contrasena = hash_contrasena(datos.contrasena),
        fecha_nacimiento=datos.fecha_nacimiento,
        telefono=datos.telefono,
        profesion=datos.profesion,
        clinica=datos.clinica
        )
        db.add(nuevo_usuario)
        db.commit()
        db.refresh(nuevo_usuario)
        token = crear_token({"sub":str(nuevo_usuario.id)})
        return TokenResponse(access_token=token,usuario=nuevo_usuario)    
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,detail="Error al registrar usuario")

@router.post("/login",response_model=TokenResponse)
def login(datos:LoginUsuario,db:Session=Depends(get_db)):
    usuario = db.query(Usuario).filter(Usuario.correo_electronico == datos.correo_electronico).first()
    if not usuario or not verificar_contrasena(datos.contrasena,usuario.contrasena):
         raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED,detail="Credenciales incorrectas")
    token = crear_token({"sub":str(usuario.id)})
    return TokenResponse(access_token=token,usuario=usuario) 

File: code/frontend_flutter/lib/models/usuario.dart
class Usuario {
  final int id;
  final String correoElectronico;
  final DateTime fechaNacimiento;
  final String? telefono;
  final String profesion;
  final String clinica;

  Usuario({
    required this.id,
    required this.correoElectronico,
    required this.fechaNacimiento,
    this.telefono,
    required this.profesion,
    required this.clinica,
  });

  factory Usuario.fromJson(Map<String, dynamic> json) {
    return Usuario(
      id: json['id'] as int,
      correoElectronico: json['correo_electronico'] as String,
      fechaNacimiento: DateTime.parse(json['fecha_nacimiento'] as String),
      telefono: json['telefono'] as String?,
      profesion: json['profesion'] as String,
      clinica: json['clinica'] as String,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'correo_electronico': correoElectronico,
      'fecha_nacimiento': fechaNacimiento.toIso8601String(),
      'telefono': telefono,
      'profesion': profesion,
      'clinica': clinica,
    };
  }
}

class LoginRequest {
  final String correoElectronico;
  final String contrasena;

  LoginRequest({required this.correoElectronico, required this.contrasena});

  Map<String, dynamic> toJson() {
    return {'correo_electronico': correoElectronico, 'contrasena': contrasena};
  }
}

class RegistroRequest {
  final String correoElectronico;
  final String contrasena;
  final DateTime fechaNacimiento;
  final String? telefono;
  final String profesion;
  final String clinica;

  RegistroRequest({
    required this.correoElectronico,
    required this.contrasena,
    required this.fechaNacimiento,
    this.telefono,
    required this.profesion,
    required this.clinica,
  });

  Map<String, dynamic> toJson() {
    return {
      'correo_electronico': correoElectronico,
      'contrasena': contrasena,
      'fecha_nacimiento': fechaNacimiento.toIso8601String().split('T').first,
      'telefono': telefono,
      'profesion': profesion,
      'clinica': clinica,
    };
  }
}

class AuthResponse {
  final String accessToken;
  final String tokenType;
  final Usuario usuario;

  AuthResponse({
    required this.accessToken,
    required this.tokenType,
    required this.usuario,
  });

  factory AuthResponse.fromJson(Map<String, dynamic> json) {
    return AuthResponse(
      accessToken: json['access_token'] as String,
      tokenType: json['token_type'] as String,
      usuario: Usuario.fromJson(json['usuario'] as Map<String, dynamic>),
    );
  }
}
